### PR TITLE
Extend and enhance networking nemesis

### PR DESCRIPTION
**Pull request by dkropachev**
https://trello.com/c/hdXhQV4q/1398-extend-and-enhance-networking-nemesis

## PR pre-checks (self review)

- [x] I followed [KISS principle](https://en.wikipedia.org/wiki/KISS_principle) and [best practices](https://docs.google.com/document/d/1jihgOKb5iGRlD8_HQ92O0JbLk1kASUoZT23i_MXFSKI)
- [x] I gave variables/functions meaningful self-explanatory names
- [x] I didn't leave commented-out/debugging code
- [x] I didn't copy-paste code
- [ ] ~~I added the relevant `backport` labels~~
- [ ] ~~New configuration option are added and documented (in `sdcm/sct_config.py`)~~
- [ ] ~~I have added tests to cover my changes (Infrastructure only - under `unit-test/` folder)~~
- [ ] ~~All new and existing unit tests passed (CI)~~
- [ ] ~~I have updated the Readme/doc folder accordingly (if needed)~~

### CODE REVIEW

**Comment by dkropachev** [comment]
https://jenkins.scylladb.com/view/staging/job/scylla-staging/job/dmitry/job/dmitry-byo-longevity/50

**Comment by dkropachev** [comment]
Successfull run:
https://jenkins.scylladb.com/view/staging/job/scylla-staging/job/dmitry/job/dmitry-byo-longevity/52

**Comment by fruch** [comment]
@bentsi @dkropachev 
1) why not keep using tcconfig ? the whole idea was that it's a wrapper ontop of iptables and others tools.
2) I would test it also when the configuration of the other network nemesis is turned on, to verify it's still working as expected, and not need to configured it to the correct interface.

**Comment by bentsi** [comment]
@fruch we can use tc for iptables too, but there are low level functionality that Dmitry probably needed, so he used iptables

File: sdcm/nemesis.py [hunk 1]
```diff
@@ -1349,6 +1349,141 @@ def disrupt_network_block(self):
...
+                'icmp-net-prohibited',
+                'icmp-host-prohibited',
+                'icmp-admin-prohibited'
+            ])
+            return f'will be rejected with {reject_with}',\
+                   f'{target_type} --reject-with {reject_with}'
+        return f'will be dropped', \
+               f'{target_type}'
+
+    def _run_commands_wait_and_cleanup(self, node, name, start_commands, cleanup_commands=None, wait_time=0):
```
Comment (bentsi) [inline sdcm/nemesis.py]: add typing annotations

File: sdcm/nemesis.py [hunk 2]
```diff
@@ -1349,6 +1349,141 @@ def disrupt_network_block(self):
...
+        Runs command/commands on target node wait and run cleanup commands
+            :param node: target node
+            :param name: Name of Nemesis for logging
+            :param start_commands: commands to run on start
+            :param cleanup_commands: commands to run to cleanup
+            :param wait_time: waiting time
+        """
+        cmd_executed = {}
+        if isinstance(start_commands, str):
+            start_commands = [start_commands]
```
Comment (bentsi) [inline sdcm/nemesis.py]: support only list type

File: sdcm/nemesis.py [hunk 3]
```diff
@@ -1349,6 +1349,141 @@ def disrupt_network_block(self):
...
+        return self._run_commands_wait_and_cleanup(
+            self.target_node,
+            name=name,
+            start_commands=f'sudo iptables -t filter -A INPUT -p tcp --dport 9160 {matching_rule} -j {pkt_action}',
+            cleanup_commands=f'sudo iptables -t filter -D INPUT -p tcp --dport 9160 {matching_rule} -j {pkt_action}',
+            wait_time=wait_time
+        )
+
+    @staticmethod
+    def _iptables_randomly_get_random_match():
```
Comment (bentsi) [inline sdcm/nemesis.py]: `_iptables_randomly_get_random_match` -> `_iptables_randomly_get_random_matching_rule`

Comment (bentsi) [reply sdcm/nemesis.py]: add a comment

File: test-cases/longevity/dk-longevity-10gb-3h.yaml [hunk 4]
```diff
@@ -0,0 +1,15 @@
+test_duration: 180
```
Comment (bentsi) [inline test-cases/longevity/dk-longevity-10gb-3h.yaml]: remove this when finished

**Comment by dkropachev** [comment]
Jenkins run on last version:
https://jenkins.scylladb.com/view/staging/job/scylla-staging/job/dmitry/job/dmitry-byo-longevity/53
Test ID:
851ddd70-61d3-4d88-92ae-119975f01b8c
In short: 
it went thru successfully for Nemesis, but in the middle of run performance dropped in twice and never recovered.

File: sdcm/nemesis.py (hunk 2 above)
Comment (dkropachev) [reply sdcm/nemesis.py]: fixed

File: sdcm/nemesis.py (hunk 3 above)
Comment (dkropachev) [reply sdcm/nemesis.py]: fixed

File: test-cases/longevity/dk-longevity-10gb-3h.yaml (hunk 4 above)
Comment (dkropachev) [reply test-cases/longevity/dk-longevity-10gb-3h.yaml]: fixed

File: sdcm/nemesis.py (hunk 1 above)
Comment (dkropachev) [reply sdcm/nemesis.py]: fixed

File: sdcm/nemesis.py [hunk 5]
```diff
@@ -760,11 +761,11 @@ def _add_drop_column_create_table(self):
                 cmd = f"DROP KEYSPACE IF EXISTS {self._add_drop_column_target_table[0]};"
                 session.execute(cmd)
                 cmd = f"CREATE KEYSPACE IF NOT EXISTS {self._add_drop_column_target_table[0]} WITH replication = " \
-                      "{'class': 'SimpleStrategy', 'replication_factor': 3};"
+                    "{'class': 'SimpleStrategy', 'replication_factor': 3};"
                 session.execute(cmd)
                 cmd = f"CREATE TABLE IF NOT EXISTS "\
-                      f"{self._add_drop_column_target_table[0]}.{self._add_drop_column_target_table[1]} " \
-                      "( col1 bigint, PRIMARY KEY(col1) );"
+                    f"{self._add_drop_column_target_table[0]}.{self._add_drop_column_target_table[1]} " \
```
Comment (bentsi) [inline sdcm/nemesis.py]: why did you change this indentation?

**Comment by bentsi** [comment]
@roydahan @fruch please review

**Comment by roydahan** [comment]
@bentsi / @dkropachev are all these nemesis part of the network longevity?
All longevity?

**Comment by dkropachev** [comment]
> @bentsi / @dkropachev are all these nemesis part of the network longevity?
> All longevity?

Since al three have network = true, it is part of NetworkMonkey and because all three methods have names starting with disrupt it, they are part of ChaosMonkey.

Our longevities has following configurations:
./test-cases/longevity/longevity-200GB-48h-network-monkey.yaml:nemesis_class_name: 'NetworkMonkey'
./test-cases/longevity/longevity-in-memory-36GB-4days.yaml:nemesis_class_name: 'ChaosMonkey'
./test-cases/longevity/longevity-large-partition-4days.yaml:nemesis_class_name: 'ChaosMonkey'
./test-cases/longevity/longevity-200GB-48h-verifier-LimitedMonkey-tls.yaml:nemesis_class_name: 'LimitedChaosMonkey'
./test-cases/longevity/longevity-1TB-7days-authorization-and-tls-ssl.yaml:nemesis_class_name: 'ChaosMonkey'
./test-cases/longevity/longevity-mv-si-4days.yaml:nemesis_class_name: 'ChaosMonkey'
./test-cases/longevity/longevity-2TB-4days-authorization-and-tls-ssl-1dis-2nondis-nemesis.yaml:nemesis_class_name: 'DisruptiveMonkey:1 NonDisruptiveMonkey:2'
./test-cases/longevity/longevity-50GB-4days-authorization-and-tls-ssl.yaml:nemesis_class_name: 'ChaosMonkey'
./test-cases/longevity/longevity-counters-4days-multidc.yaml:nemesis_class_name: 'ChaosMonkey'
./test-cases/longevity/longevity-alternator-3h.yaml:nemesis_class_name: NoOpMonkey # ChaosMonkey
./test-cases/longevity/longevity-10gb-3h.yaml:nemesis_class_name: 'ChaosMonkey'
./test-cases/longevity/longevity-100GB-48h-cloud-CloudLimitedChaosMonkey-tls.yaml:nemesis_class_name: 'LimitedChaosMonkey'
./test-cases/longevity/longevity-encryption-at-rest-200GB-6h.yaml:nemesis_class_name: 'LimitedChaosMonkey'
./test-cases/longevity/longevity-5000-tables.yaml:nemesis_class_name: NoOpMonkey
./test-cases/longevity/longevity-multi-keyspaces.yaml:nemesis_class_name: 'ChaosMonkey'
./test-cases/longevity/longevity-100gb-4h.yaml:nemesis_class_name: 'ChaosMonkey'

So, it is part of nearly all longevities except:
./test-cases/longevity/longevity-100GB-48h-cloud-CloudLimitedChaosMonkey-tls.yaml:nemesis_class_name: 'LimitedChaosMonkey'
./test-cases/longevity/longevity-encryption-at-rest-200GB-6h.yaml:nemesis_class_name: 'LimitedChaosMonkey'
./test-cases/longevity/longevity-200GB-48h-verifier-LimitedMonkey-tls.yaml:nemesis_class_name: 'LimitedChaosMonkey'
./test-cases/longevity/longevity-alternator-3h.yaml:nemesis_class_name: NoOpMonkey # ChaosMonkey
